torch/distributed/rpc.py

```diff
@@ -0,0 +1,137 @@
+from . import invoke_rpc
+from . import ProcessGroupAgent
+
+import array
+import torch
+
+_agent = None
+
+def _collect_worker_names(name, group):
+    from . import all_gather
+    from . import get_world_size
+
+    # collect name length
+    ws = get_world_size(group)
+    encoded_name = list(name.encode('utf-8'))
+    name_len = len(encoded_name)
+    len_input = torch.ones(1, dtype=torch.int64) * name_len
+    len_outputs = [torch.empty(1, dtype=torch.int64) for _ in range(ws)]
+    all_gather(len_outputs, len_input, group=group)
+
+    # collect name value
+    max_len = torch.stack(len_outputs).max().item()
+    name_input = torch.empty(max_len, dtype=torch.uint8)
+    name_input[:name_len] = torch.tensor(encoded_name, dtype=torch.uint8)
+    name_outputs = [torch.empty(max_len, dtype=torch.uint8) for _ in range(ws)]
+    all_gather(name_outputs, name_input, group=group)
+
+    names = []
+    for i in range(ws):
+        name_tensor = name_outputs[i][:len_outputs[i]]
+        names.append(array.array('B', name_tensor.tolist()).tobytes().decode('utf-8'))
+
+    return names
+
+
+def join_rpc():
+    r"""
+    Block until all local and remote RPC processes reach this method, process
+    (send and receive) all pending messages, and then destroy local RPC agent.
+    Every RPC process must call this method before exit.
```

Comment:
If every RPC process must call this before exit, should we provide a context manager for `init_rpc` and `join_rpc`?

Reply:
good point, let me add one.